### PR TITLE
Updates Kubeproxy validators to allow Windows 'kernelspace' mode.

### LINKED ISSUE #56522: Kubeproxy's Windows proxy modes don't work.
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**: When running `kubeproxy.exe`, and passing the `--proxy-mode=kernelspace` flag which was introduced in [PR #51064 ](https://github.com/kubernetes/kubernetes/pull/51064), the following error occurs:

```
error: KubeProxyConfiguration.Mode.ProxyMode: Invalid value: "kernelspace": must be 
userspace,iptables,ipvs or blank (blank means the best-available proxy (currently iptables)
```

**What you expected to happen**: The proxy should work as it used to, supporting this mode.

**How to reproduce it (as minimally and precisely as possible)**: 

    ./kube-proxy.exe --proxy-mode=kernelspace

**Anything else we need to know?**: This was introduced after the validation was refactored in [this PR](https://github.com/kubernetes/kubernetes/pull/49087) which regressed validation for `--proxy-mode=kernelspace`. IPVS validation (and thus support) was later added [here](https://github.com/kubernetes/kubernetes/pull/53780), but we need another mode for Windows.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.10.0-alpha.0.684+51033c4dec6e00
- Cloud provider or hardware configuration: n/a
- OS (e.g. from `/etc/os-release`):  Windows Server Core 1709
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by gkudra-msft**
**What this PR does / why we need it**: 
Allows necessary `--proxy-mode` parameter in Kubeproxy, so that it can proceed as usual on Windows.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: 
Fixes #56522

```release-note
NONE
```

### CODE REVIEW

**Comment by gkudra-msft** [comment]
I think this should actually go into `release-1.9` but I will defer to the reviewer(s).

/assign thockin 
CC @madhanrm

File: pkg/proxy/apis/kubeproxyconfig/validation/validation.go
Comment (thockin) [inline pkg/proxy/apis/kubeproxyconfig/validation/validation.go]: This logic really needs to validate based on GOOS.  It doesn't make sense to allow options from the wrong OS.

Comment (gkudra-msft) [reply pkg/proxy/apis/kubeproxyconfig/validation/validation.go]: Agreed and added. Not sure if you noticed my comment, but should this go into `release-1.9` instead? This is a full-stop blocker for deploying Windows nodes so we'd want this in ASAP.

**Comment by gkudra-msft** [comment]
Fixed a small Go style mishap. 

/retest

**Comment by gkudra-msft** [comment]
Ping @thockin.

File: pkg/proxy/apis/kubeproxyconfig/validation/validation.go
Comment (thockin) [inline pkg/proxy/apis/kubeproxyconfig/validation/validation.go]: This still isn't right - ipvs and iptables are linux only.  This should just be broken to two different functions.

```
if runtime.GOOS == "windows" {
  return validateProxyModeWindows(mode, fldPath)
} else {
   return validateProxyModeLinux(mode, fldPath)
}
```

wherein each function does just what is allowed for that OS

**Comment by gkudra-msft** [comment]
I hope this is fine? @thockin 

Also, any chance this could get into `1.9.0-beta.2` today?

**Review by bsteciuk — APPROVED** [review]
I tested out these changes and can confirm that they do fix the problem.  I also agree with @gkudra-msft , if possible, it'd be great to get this in so Windows users don't need to build their own kube-proxy with this fix,

**Comment by PatrickLang** [comment]
ping @JiangtianLi - can your team help with a review? we need this merged ASAP because Windows Server 1709 is broken without this. The window for 1.9 cherry picks closes today

File: pkg/proxy/apis/kubeproxyconfig/validation/validation.go
Comment (dcbw) [inline pkg/proxy/apis/kubeproxyconfig/validation/validation.go]: Instead of checking GOOS, could we perhaps have separate validation_windows.go and validation_linux.go files with +build directives instead?  That seems to be a normal pattern, at least more normal than checking GOOS.

File: pkg/proxy/apis/kubeproxyconfig/validation/validation.go
Comment (thockin) [inline pkg/proxy/apis/kubeproxyconfig/validation/validation.go]: I won't block this PR, but this is brittle - listing modes twice.  Better would be:

```
var modesLinux = sets.NewString(
    string(kubeproxyconfig.ProxyModeUserspace),
    string(kubeproxyconfig.ProxyModeIPTables),
    string(kubeproxyconfig.ProxyModeIPVS)
)
func validateProxyModeLinux(mode, ...) {
    if mode == "" || modesLinux.Has(mode) {
        return nil
    }
    return field.ErrorList{field.Invalid(fldPath.Child("ProxyMode"), string(mode), errMsg)}
}
```

or something - the key being that the list of modes is spelled out exactly one time.

**Review by thockin — COMMENTED** [review]
Will approve,  but this can be cleaned up.

**Comment by JiangtianLi** [comment]
LGTM

**Comment by thockin** [comment]
please follow-up with cleanups.

**Review by JiangtianLi — APPROVED** [review]

**Comment by jberkus** [comment]
Can someone add this to the 1.9 milestone and approve it?

**Comment by gkudra-msft** [comment]
Test errors are unrelated and timeouts.

/retest

File: pkg/proxy/apis/kubeproxyconfig/validation/validation.go
Comment (feiskyer) [reply pkg/proxy/apis/kubeproxyconfig/validation/validation.go]: File a PR to fix this #57127

**Pull request merged**
